server/http/httpclient/httpclient.go

```diff
@@ -0,0 +1,125 @@
+package httpclient
+
+import (
+	"errors"
+	"flag"
+	"net"
+	"net/http"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/net/publicsuffix"
+)
+
+var allowLocalhost = flag.Bool("httpclient.allow_localhost", false, "Allow HTTP requests to localhost")
+
+const maxHTTPTimeout = 60 * time.Minute
+
+type dialerControl = func(network, address string, conn syscall.RawConn) error
+
+func blockingDialerControl(allowed []*net.IPNet) dialerControl {
+	return func(network, address string, conn syscall.RawConn) error {
+		host, _, err := net.SplitHostPort(address)
+		if err != nil {
+			return err
+		}
+		ip := net.ParseIP(host)
+		for _, ipNet := range allowed {
+			if ipNet.Contains(ip) {
+				return nil
+			}
+		}
+		if ip.IsLoopback() || ip.IsPrivate() || ip.IsLinkLocalUnicast() || ip.IsLinkLocalMulticast() {
+			log.Infof("Blocked Fetch for address %s", address)
+			return errors.New("IP address not allowed")
+		}
+		return nil
+	}
+}
+
+func NewClientWithPrivateIPNets(timeout time.Duration, allowedPrivateIPNets []*net.IPNet) *http.Client {
+	dialerTimeout := timeout
```

Comment:
Maybe call this `NewWithAllowedPrivateIPs`? `WithPrivateIPNets` doesn't convey what this list does. 

Since this client is used in multiple places and likely to be used in others, it's worth adding some documentation to these functions to explain the behavior of this package.